### PR TITLE
STORE 'file' scheme loader: fix try_decode_params() to check ambiguity

### PR DESCRIPTION
**Pull request by levitte**
The way try_decode_params works in raw more, it would take the first ASN1
that could decode and return a STORE_INFO with the resulting EVP_PKEY.
This change has it go through all the matching ASN1 methods and properly
check if there's more than one match, i.e. an ambiguity.

### CODE REVIEW

File: crypto/store/loader_file.c [hunk 1]
```diff
@@ -480,35 +480,52 @@ static OSSL_STORE_INFO *try_decode_params(const char *pem_name,
...
             && ameth->param_decode(pkey, &blob, len))
             ok = 1;
     } else {
         int i;
+        EVP_PKEY *tmp_pkey = NULL;
 
         for (i = 0; i < EVP_PKEY_asn1_get_count(); i++) {
             const unsigned char *tmp_blob = blob;
 
+            if (tmp_pkey == NULL && (tmp_pkey = EVP_PKEY_new()) == NULL) {
```
Comment (kaduk) [inline crypto/store/loader_file.c]: I'm not sure that I see a way for tmp_pkey to be non-NULL here, but it's probably not harmful to keep the check.

**Review by kaduk — APPROVED** [review]
It's a little strange to have such different code structures for the same functionality here vs. try_decode_PrivateKey, but that's a side note.

File: crypto/store/loader_file.c [hunk 2]
```diff
@@ -480,35 +480,52 @@ static OSSL_STORE_INFO *try_decode_params(const char *pem_name,
...
-                && ameth->param_decode(pkey, &tmp_blob, len)) {
+                && ameth->param_decode(tmp_pkey, &tmp_blob, len)) {
+                if (pkey != NULL)
+                    EVP_PKEY_free(tmp_pkey);
+                else
+                    pkey = tmp_pkey;
+                tmp_pkey = NULL;
                 (*matchcount)++;
-                ok = 1;
                 break;
```
Comment (kaduk) [inline crypto/store/loader_file.c]: We should remove this break

**Review by kaduk — CHANGES_REQUESTED** [review]

File: crypto/store/loader_file.c (hunk 1 above)
Comment (levitte) [inline crypto/store/loader_file.c]: `tmp_pkey` is NULL on the first iteration, and then, when there's a successful param_decode, `tmp_pkey` gets reset to NULL, which affects the next iteration (I must just remove the `break` that you pointed out).

Comment (kaduk) [inline crypto/store/loader_file.c]: Right, so it's NULL the first time and gets reset to NULL before the second, third, etc. times -- it's always NULL.

**Comment by levitte** [comment]
> It's a little strange to have such different code structures for the same functionality here vs. try_decode_PrivateKey, but that's a side note.

That's because there isn't a generic `d2i_Params` like there is a generic `d2i_PrivateKey`...  otherwise, they would have been very similar.

File: crypto/store/loader_file.c (hunk 2 above)
Comment (levitte) [inline crypto/store/loader_file.c]: Gone now

File: crypto/store/loader_file.c (hunk 1 above)
Comment (levitte) [inline crypto/store/loader_file.c]: Not always, only for those where param_decode succeeds.  For those that fail, the previously allocated one is still used.

Comment (kaduk) [inline crypto/store/loader_file.c]: Ah, the unsuccessful decode leaves it around, I see now.

Comment (kaduk) [inline crypto/store/loader_file.c]: And we really need to stop racing when commenting :)

**Review by kaduk — APPROVED** [review]

File: crypto/store/loader_file.c (hunk 1 above)
Comment (levitte) [inline crypto/store/loader_file.c]: Haha :-)

**Comment by levitte** [comment]
Merged.  11d66064f36e6968faffb48a2cfd58cbe37eff0c

Thanks for the review!

**Pull request closed without merging**